PROVIDER.md

```diff
@@ -1,168 +1,201 @@
-type converter struct {
-	conf *i2gw.ProviderConf
-
-	featureParsers []i2gw.FeatureParser
-	implementationSpecificOptions i2gw.ProviderImplementationSpecificOptions
-}
-
-// newConverter returns an ingress-nginx converter instance.
-func newConverter(conf *i2gw.ProviderConf) *converter {
-	return &converter{
-		conf: conf,
-		featureParsers: []i2gw.FeatureParser{
-			// The list of feature parsers comes here.
-		},
-		implementationSpecificOptions: i2gw.ProviderImplementationSpecificOptions{
-			// The list of the implementationSpecific ingress fields options comes here.
-		},
-	}
-}
-```
-4. Create a new struct named after the provider you are implementing. This struct should embed the previous 2 structs 
-you created.
-```go
-package examplegateway
-
-import (
-	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw"
-)
-
-// Provider implements the i2gw.Provider interface.
-type Provider struct {
-	conf *i2gw.ProviderConf
-
-	*resourceReader
-	*converter
-}
-
-// NewProvider constructs and returns the example-gateway implementation of i2gw.Provider.
-func NewProvider(conf *i2gw.ProviderConf) i2gw.Provider {
-	return &Provider{
-		conf:           conf,
-		resourceReader: newResourceReader(conf),
-		converter:      newConverter(conf),
-	}
-}
-```
-5. Add the new provider to `i2gw.ProviderConstructorByName`.
-```go
-package examplegateway
-
-import (
-	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw"
-)
-
-// The Name of the provider.
-const Name = "example-gateway-provider"
-
-func init() {
-	i2gw.ProviderConstructorByName[Name] = NewProvider
-}
-```
-6. Import the new package at `cmd/print`.
-```go
-package cmd
-
-import (
-	// Call init function for the providers
-	_ "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/ingressnginx"
-	_ "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/examplegateway"
-)
-```
+   `resource_reader.go`.
+
+   ```go
+   package examplegateway
+
+   import (
+       "context"
+
+       "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/common"
+   )
+
+   // converter implements the i2gw.CustomResourceReader interface.
+   type resourceReader struct {
+       conf *i2gw.ProviderConf
+   }
+
+   // newResourceReader returns a resourceReader instance.
+   func newResourceReader(conf *i2gw.ProviderConf) *resourceReader {
+       return &resourceReader{
+           conf: conf,
+       }
+   }
+
+   func (r *resourceReader) readResourcesFromCluster(ctx context.Context) (*storage, error) {
+       // read example-gateway related resources from the cluster.
+       return nil, nil
+   }
+
+   func (r *resourceReader) readResourcesFromFile(ctx context.Context, filename string) (*storage, error) {
+       // read example-gateway related resources from the file.
+       return nil, nil
+   }
+   ```
+
+   These methods are used by providers to read and store additional resources they may need during conversion.
+
+3. Create a struct named `storage` in a file named `storage.go`:
```

Comment:
maybe add instructions on what should be stored in storage? i.e  ingresses, CRDs or essentially more things that the provider need for the conversion